docs/upgrade.md

```diff
@@ -3,5 +3,263 @@ title: Upgrade guide
+| **Logger - log sampling**              | Changed implementation of [log sampling](./core/logger.md#sampling-logs) to dynamically switch log level to `DEBUG` on a percentage of requests. | -                    |
+| **Logger - Custom Log Formatter**      | Updated [custom log formatter](#custom-log-formatter) to include additional persistent attributes along with standard structured keys.           | Yes                  |
+| **Logger & Tracer - helper functions** | Removed deprecated `createLogger` and `createTracer` helper functions in favor of direct instantiation.                                          | Yes                  |
+
+### First steps
+
+Before you start, we suggest making a copy of your current working project or create a new git branch.
+
+1. Upgrade Node.js to v16 or higher, Node.js v20 is recommended.
+2. Ensure that you have the latest Powertools for AWS Lambda (TypeScript) version via [Lambda Layer](./index.md#lambda-layer) or npm.
+3. Review the following sections to confirm whether they apply to your codebase.
+
+## ES Modules support
+
+Starting with v2, Powertools for AWS Lambda (TypeScript) supports ES Modules. This means that you can now import the package using the `import` syntax instead of the `require` syntax. This is especially useful if you want to leverage features like top-level `await` in your Lambda function to run asynchronous code during the initialization phase.
+
+```typescript
+import { getSecret } from '@aws-lambda-powertools/parameters/secrets';
+
+// This code will run during the initialization phase of your Lambda function
+const myApiKey = await getSecret('my-api-key', { transform: 'json' });
+
+export const handler = async () => {
+    // ...
+};
+```
+
+With this change, you can also apply tree-shaking to your function bundle to reduce its size. As part of this release we have made changes to the package and its exports to better support this feature, and we remain committed to improving this in the future based on your feedback.
+
+While we recommend using ES Modules, we understand that this change might not be possible for everyone. If you're unable to use ES Modules, you can continue to use the `require` syntax to import the package. Powertools for AWS Lambda (TypeScript) will continue to support this syntax by shipping CommonJS modules alongside ES Modules.
+
+In some cases, even when opting for ES Modules, you might still need to use the `require` syntax to import a package. For example, if you're using a package that doesn't support ES Modules, or if one of your transitive dependencies is using the `require` syntax like it's the case for `@aws-lambda-powertools/tracer` which relies on the AWS X-Ray SDK for Node.js. In these cases, you can still use ES Modules for the rest of your codebase and set a special build flag to tell your bundler to inject a banner at the top of the file to use the `require` syntax for the specific package.
```

Comment:
```suggestion
### Unable to use ESM?

!!! note "We recommend using ESM for the best experience _(top-level await, smaller bundle size etc.)_."

If you're unable to use ESM, you can still use the `require` syntax to import the package. We will continue to support it by shipping CommonJS modules alongside ESM.

You might still need the `require` syntax when using a dependency or a transitive dependency that doesn't support ESM. For example, Tracer _(`@aws-lambda-powertools/tracer`)_ relies on the AWS X-Ray SDK for Node.js which uses `require`.

When that happens, you can instruct your bundler to use the `require` syntax for specific dependencies while using ESM for everything else. This is commonly known as [polyfill](https://developer.mozilla.org/en-US/docs/Glossary/Polyfill){target="_blank"}. 

Here is an example using `esbuild` bundler.
```

It'll look like this at the end. Why have I done this?

* I’m using a separate section to lure people who can’t use ESM
* A separate section will also save discussions about this topic thanks to deep link
* I’m breaking that terse but useful information in three paragraphs so they can read, memorize, digest, and move on to the next bit
* I’m sending them at the end to MDN to learn more about the term that’s been there for a while so they can become experts too

tip: lead with the answer / use case, learn about `sentence fragmentation`, join them, then try cutting to have the least amount of sentences while keeping the core of your message.

<img width="1561" alt="image" src="https://github.com/aws-powertools/powertools-lambda-typescript/assets/3340292/973909c5-e042-4f77-b011-cafda363649d">